### ISSUE TITLE
2.2.0 installs wrong .pc file

### ISSUE THREAD

**Issue by yurivict** [openkim/kim-api#57]
```libkim-api.pc``` contains wrong paths:
```
$ pkg-config --cflags --libs libkim-api
-I/usr/local/libdata/pkgconfig/../../include/kim-api -I/usr/local/libdata/pkgconfig/../kim-api/mod -L/usr/local/libdata/pkgconfig/../ -Wl,-rpath,/usr/local/libdata/pkgconfig/../ -lkim-api 
```

Problems:
1. Path ```/usr/local/libdata/pkgconfig/../kim-api/mod``` doesn't exist.
2. Path ```/usr/local/libdata/pkgconfig/../``` isn't a library directory

Please note that the path where pkgconfig script is installed is unrelated to C++ header/library paths.

Also please do not use ```/..``` in the paths, please use absolute paths. These only bloat command lines and cause confusion.

**Comment by ellio167** [comment]
Thanks for the report.

The relative paths are necessary to make the kim-api package relocatable.  The `pkg-config` command is specifically designed to support/encourage this behavior through the `pcfiledir` variable (see the man page).

The kim-api cmake build scripts know where the `libkim-api.pc` file is installed (in the `${CMAKE_INSTALL_LIBDIR}/pkgconfig` directory) and set its variables appropriately.

If a distribution decides to place this file elsewhere, then the packaging scripts should ensure the `libkim-api.pc` file is update as necessary.

I see that for autotools there is a standard `--with-pkgconfigdir` that will override the location.  I would be happy to implement a setting like this.  Does anyone know, is there a "standard" CMake spelling `-DCMAKE_INSTALL_PKGCONFIGDIR` or something?

**Comment by ellio167** [comment]
@junghans

**Comment by yurivict** [comment]
> If a distribution decides to place this file elsewhere, then the packaging scripts should ensure the libkim-api.pc file is update as necessary.

No, this is the responsibility of the project to provide a valid .pc file. Distribution maintainers are not responsible to correct paths in installed files.

cmake has all required paths in its variables to fill .pc file correctly.

Location of .pc file is irrelevant to what paths it provides.

**Comment by ellio167** [comment]
> Location of .pc file is irrelevant to the paths it provides.

I guess we'll have to disagree on this point.  I'm following the example from the `pkg-config` manpage:

```
       pcfiledir
              The installed location of the .pc file.  This  can  be  used  to
              query  the location of the .pc file for a particular module, but
              it can also be used to make .pc files relocatable. For instance:
              prefix=${pcfiledir}/../..
              exec_prefix=${prefix}
              libdir=${exec_prefix}/lib
              includedir=${prefix}/include
```
which to my mind, shows that paths relative to where the .pc file is located is an fully supported use-case.

**Comment by yurivict** [comment]
> which to my mind, shows that paths relative to where the .pc file is located is an fully supported use-case.

If this is what is the recommended way is, this means that you have wrong number of ```..``` in the ```pc```.

**Comment by yurivict** [comment]
Particularly, it installs .pc file with
```
moddir=${pcfiledir}/../kim-api/mod
```
when it should be
```
moddir=${pcfiledir}/../../include/kim-api/mod
```
to match where ```kim-api``` is installed.

**Comment by ellio167** [comment]
Right; as I understand it this is because the packaging tools are moving the .pc file from the location it is installed to by cmake to the location that FreeBSD wants it.  I don't see anywhere that this is being done explicitly in your [package files](https://svnweb.freebsd.org/ports?view=revision&revision=555727), but the [build log](http://beefy6.nyi.freebsd.org/data/121amd64-default/555803/logs/kim-api-2.2.0.log) show it installed under `lib`:
`-- Installing: /wrkdirs/usr/ports/science/kim-api/work/stage/usr/local/lib/pkgconfig/libkim-api.pc`
whereas the [/head/science/kim-api/pkg-plist](https://svnweb.freebsd.org/ports/head/science/kim-api/pkg-plist?view=markup&pathrev=555727) file shows in under `libdata`:
`138 | libdata/pkgconfig/libkim-api.pc`

If the .pc file were under `lib` it would work correctly, I think.

**Comment by yurivict** [comment]
No, ```.pc``` file is installed into a location exactly following what the manpage says that you cited above (2 directories below ```$prefix```).

The paths that you have in the ```.pc``` file are simply wrong, not matching the files that you install.

**Comment by ellio167** [comment]
I've triple checked.  The kim-api cmake build installs to `${PREFIX}/lib/pkgconfig/libkim-api.pc`.  Thus, the contents of that file are correct.  In particular the line `libdir=${pcfiledir}/../` resolves to `${PREFIX}/lib/` as desired.  The file is correct.

However, the FreeBSD package somehow has the file in `/usr/local/libdata/pkgconfig/libkim-api.pc` which breaks the file's contents.

I see that the man page says:
```
--define-prefix
       --dont-define-prefix  These  options  control whether pkg-config overrides the value of the variable prefix in each .pc file. With
       --define-prefix, pkg-config uses the installed location of the .pc file to determine  the  prefix.  --dont-define-prefix  prevents
      this behavior. The default is usually --define-prefix.

       When  this feature is enabled and a .pc file is found in a directory named pkgconfig, the prefix for that package is assumed to be
       the grandparent of the directory where the file was found, and the prefix variable is overridden for that file accordingly.

       If the value of a variable in a .pc file begins with the original, non-overridden, value of the prefix variable, then the overrid-
       den value of prefix is used instead. This allows the feature to work even when the variables have been expanded in the .pc file.
```
But, the prefix variable is not used within the `libkim-api.pc` file, so I don't see why it should be required to write it as `${pcfiledir}/../../lib` when `${pcfiledir}/../` is equivalent.

**Comment by yurivict** [comment]
> I've triple checked. The kim-api cmake build installs to ${PREFIX}/lib/pkgconfig/libkim-api.pc. Thus, the contents of that file are correct. In particular the line libdir=${pcfiledir}/../ resolves to ${PREFIX}/lib/ as desired. The file is correct.


The manpage you mentioned instructs you that ```${pcfiledir}/../..``` is equal to ```${pfrefix}```. You can't make any further assumptions. Particularly, you can't assume that ```${pcfiledir}/..```=```${prefix}/lib```.

**Comment by ellio167** [comment]
Regardless, the discussion here is not leading us to a practical solution.

Would it be acceptable to you to use a cmake argument for this?  Instead of just this
`USES=           cmake:noninja fortran tar:txz`
in /head/science/kim-api/Makefile it would be
`USES=           cmake:noninja fortran tar:txz`
`CMAKE_ARGS= -DCMAKE_INSTALL_LIBDATADIR=libdata`
?

**Comment by yurivict** [comment]
> Regardless, the discussion here is not leading us to a practical solution.

Yes it does. You have clear mistakes in the .pc file.

> Would it be acceptable to you to use a cmake argument for this? Instead of just this

This can be done as well, but fixing the ```.pc``` file would also do.

**Comment by ellio167** [comment]
> The manpage you mentioned instructs you that ${pcfiledir}/../.. is equal to ${pfrefix}. You can't make any further assumptions. Particularly, you can't assume that ${pcfiledir}/..=${prefix}/lib.

I don't see why you insist on this.  The man page says `For instance:`, it is an example not a requirement.
```
       pcfiledir
              The installed location of the .pc file.  This  can  be  used  to
              query  the location of the .pc file for a particular module, but
              it can also be used to make .pc files relocatable. For instance:
              prefix=${pcfiledir}/../..
              exec_prefix=${prefix}
              libdir=${exec_prefix}/lib
              includedir=${prefix}/include
```

**Comment by yurivict** [comment]
> I don't see why you insist on this. The man page says For instance:, it is an example not a requirement.

If you would be following what the manpage suggests this would solve the problem.
Particularly, first compute ```prefix=${pcfiledir}/../..```, and then derive all further values from ```prefix```.

**Comment by yurivict** [comment]
The current .pc file has this section:
```
prefix=${pcfiledir}/../../
bindir=${pcfiledir}/../../bin
libexecdir=${pcfiledir}/../../libexec
sysconfdir=${pcfiledir}/../../etc
libdir=${pcfiledir}/../
includedir=${pcfiledir}/../../include/kim-api
moddir=${pcfiledir}/../kim-api/mod
docdir=${pcfiledir}/../../share/doc/kim-api
```
This is *not* how the manpage recommends to do it.

It should read:
```
prefix=${pcfiledir}/../../
bindir=${prefix}/bin
libexecdir=${prefix}/libexec
sysconfdir=${prefix}/etc
libdir=${prefix}/lib
includedir=${prefix}/include/kim-api
moddir=${prefix}/lib/kim-api/mod
docdir=${prefix}/share/doc/kim-api
```

**Comment by ellio167** [comment]
OK I see how this can be the preferred format.  @yurivict thanks for your attention to this; A fix will be included in v2.2.1
